# Numpy Fundamentals

# What is Numpy

* Numpy stands for `Numerical` `Python`.
* Numpy is the fundamental package for scientific computing in Python.
* Numpy is an open source `Python` library used to working with `array`.
* Numpy is partially written in `Python`, but most of the parts written in `C` or `C++` for fast computation .
* Numpy is creared in `2005` by `Travis Oliphant`.
* It also has function for working in domain of `linear algebra`, `fourier transform` and `matrices`.
* Numpy arrays are `Homogeneous` unlike lists are `Hetrogeneous`.
* The source code for NumPy is located at this github repository https://github.com/numpy/numpy.

### Use of Numpy

* In Python we have list but they are slow, becuase lists are not directly store in the memory only it's reference stored in memory.
* While Numpy arrays stored at one continious place in memory.
* This behavior is called `locality of reference` in computer science.
* The array object in Python is called `ndarray` it provides lots of fuction which make working with `ndarray` is very easy.
* Arrays are widely used in the field of `Data Science` where speed and resources are very important.

# Creating an Array

In [1]:
import numpy as np

In [2]:
arr = np.array([1,2,3,4])
print(arr)

[1 2 3 4]


In [3]:
print(np.__version__)

1.21.5


In [4]:
arr = np.array([1,2,3,4])
print(arr)
print(type(arr))

[1 2 3 4]
<class 'numpy.ndarray'>


* Use a tuple to create a Numpy array
* we can pass any type of array type object to `array()` method and it will converted to `ndarray`.


In [5]:
tup_arr = np.array((1,2,3,4,5))
print(tup_arr)
print(type(tup_arr))

[1 2 3 4 5]
<class 'numpy.ndarray'>


In [6]:
arr = np.arange(1, 11, 2) # Same as range function in Python.
arr

array([1, 3, 5, 7, 9])

In [7]:
arr = np.arange(1, 11).reshape(2, 5) # reshape convert one aray into another shape
arr
# It will work when product of two number should be equal to the numbers that present in the array

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10]])

In [8]:
# arr = np.arange(1, 13).reshape(5, 5)
# arr
# This will through error

In [9]:
arr = np.arange(16).reshape(8,2)
arr

array([[ 0,  1],
       [ 2,  3],
       [ 4,  5],
       [ 6,  7],
       [ 8,  9],
       [10, 11],
       [12, 13],
       [14, 15]])

In [10]:
arr = np.array([1,2,3,4], dtype=float) # This a type conversion
arr

array([1., 2., 3., 4.])

In [11]:
arr = np.array([1,2,3,4], dtype=bool) # This a type conversion
arr

array([ True,  True,  True,  True])

In [12]:
arr = np.array([1,2,3,4], dtype=complex) # This a type conversion
arr

array([1.+0.j, 2.+0.j, 3.+0.j, 4.+0.j])

In [13]:
np.ones((3,4)) # By default numpy array elemsnts are float

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [14]:
np.zeros((3,4))

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [15]:
np.random.random((3,4))

array([[0.15366495, 0.07797812, 0.53122692, 0.66288169],
       [0.1883758 , 0.91996651, 0.32236377, 0.94106944],
       [0.86492024, 0.8000769 , 0.97021215, 0.79824574]])

In [16]:
# Linear Space
np.linspace(-10, 10, 10) # (Lower_range, upper_range, number_of_items)
# generate equally distanced points in a given range or we can say
# generate linearly separable point in a given range

array([-10.        ,  -7.77777778,  -5.55555556,  -3.33333333,
        -1.11111111,   1.11111111,   3.33333333,   5.55555556,
         7.77777778,  10.        ])

In [17]:
np.linspace(-10, 10, 10, dtype=int) # Can be change data type of elements

array([-10,  -8,  -6,  -4,  -2,   1,   3,   5,   7,  10])

In [18]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

# Array Attributes

In [19]:
a1 = np.arange(10)
a2 = np.arange(12, dtype=float).reshape(3,4)
a3 = np.arange(8).reshape(2,2,2)

#### ndim

In [20]:
print(a1.ndim)
print(a2.ndim)
print(a3.ndim)
print(a2)

1
2
3
[[ 0.  1.  2.  3.]
 [ 4.  5.  6.  7.]
 [ 8.  9. 10. 11.]]


In [21]:
zero_d_arr = np.array(1)
one_d_arr = np.arange(10)
two_d_arr = np.arange(1,11).reshape(2,5)
three_d_arr= np.arange(8).reshape(2,2,2)

In [22]:
print(f"Dimension of 0-D Array: {zero_d_arr.ndim}")
print(f"Dimension of 1-D Array: {one_d_arr.ndim}")
print(f"Dimension of 2-D Array: {two_d_arr.ndim}")
print(f"Dimension of 3-D Array: {three_d_arr.ndim}")

Dimension of 0-D Array: 0
Dimension of 1-D Array: 1
Dimension of 2-D Array: 2
Dimension of 3-D Array: 3


In [23]:
three_d_arr = np.array([   [[1,2,3],[4,5,6]],   [[7,8,9],[10,11,12]]   ])
print(three_d_arr)
print(three_d_arr.ndim)

[[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]]
3


#### shape

In [24]:
print(a1.shape)
print(a2.shape)
print(a3.shape) # (No. of 2-D arrays, No. of rows, No. of columns)

(10,)
(3, 4)
(2, 2, 2)


#### size

In [25]:
print(a1.size)
print(a2.size)
print(a3.size)

10
12
8


#### itemsize

In [26]:
# By default use int64
# int32 = 4 bytes
# int64 = 8 bytes
print(a1.itemsize)
print(a2.itemsize)
print(a3.itemsize)

4
8
4


#### dtype

In [27]:
print(a1.dtype)
print(a2.dtype)
print(a3.dtype)

int32
float64
int32


# Changing Data Types

#### astype

In [28]:
a3.dtype

dtype('int32')

In [29]:
a3.astype(np.int64) # Used for memory optimization

array([[[0, 1],
        [2, 3]],

       [[4, 5],
        [6, 7]]], dtype=int64)

# Array Operations

In [30]:
a2 = np.arange(12).reshape(3,4)
a3 = np.arange(12,24).reshape(3,4)

In [31]:
# Scalar Operations
a2 * 2 # whereas 2 is scalar
a2 + 2
a2 - 2
a2 / 2
a2**2
# Arithmetic

array([[  0,   1,   4,   9],
       [ 16,  25,  36,  49],
       [ 64,  81, 100, 121]], dtype=int32)

In [32]:
# Relational operators
print(a2)
print(a2 >= 5) # Itemwise comparision

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
[[False False False False]
 [False  True  True  True]
 [ True  True  True  True]]


In [33]:
# Vector Operations
# Shape msut be same

print(a2, "\n")
print(a3, "\n")
print(a2 + a3)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]] 

[[12 13 14 15]
 [16 17 18 19]
 [20 21 22 23]] 

[[12 14 16 18]
 [20 22 24 26]
 [28 30 32 34]]


# Array Functions

In [34]:
a1 = np.random.random((3,3))
a1 = np.round(a1*100)
print(a1)

[[73. 18. 96.]
 [13. 71. 81.]
 [50. 10.  1.]]


In [35]:
# max/min/sum/prod
np.max(a1)

96.0

In [36]:
# 0 -> Column
# 1 -> row
np.max(a1, axis=1) # gives maximum of each row

array([96., 81., 50.])

In [37]:
np.max(a1, axis=0) # gives maximum of each column

array([73., 71., 96.])

In [38]:
np.min(a1)

1.0

In [39]:
np.sum(a1)

413.0

In [40]:
np.prod(a1)

4715451936000.0

In [41]:
# mean/median/std/var
np.mean(a1)

45.888888888888886

In [42]:
np.mean(a1, axis=1)

array([62.33333333, 55.        , 20.33333333])

In [43]:
np.median(a1)

50.0

In [44]:
np.std(a1)

33.81138678822875

In [45]:
np.var(a1)

1143.20987654321

In [46]:
# Trignometric Fucntions

np.sin(a1)

array([[-0.67677196, -0.75098725,  0.98358775],
       [ 0.42016704,  0.95105465, -0.62988799],
       [-0.26237485, -0.54402111,  0.84147098]])

In [47]:
np.cos(a1)

array([[-0.73619272,  0.66031671, -0.18043045],
       [ 0.90744678, -0.30902273,  0.77668598],
       [ 0.96496603, -0.83907153,  0.54030231]])

In [48]:
np.tan(a1)

array([[ 0.9192864 , -1.13731371, -5.45134011],
       [ 0.46302113, -3.0776204 , -0.81099442],
       [-0.27190061,  0.64836083,  1.55740772]])

In [49]:
# dot product
a2 = np.arange(12).reshape(3,4)
a3 = np.arange(12,24).reshape(4,3)

* `(3, 4)` `(4, 3)` first array column must be equal to second array row
* And the result would be first array row x second array column


In [50]:
np.dot(a2, a3)

array([[114, 120, 126],
       [378, 400, 422],
       [642, 680, 718]])

In [51]:
# log / exponents
np.log(a1)

array([[4.29045944, 2.89037176, 4.56434819],
       [2.56494936, 4.26267988, 4.39444915],
       [3.91202301, 2.30258509, 0.        ]])

In [52]:
np.exp(a1)

array([[5.05239363e+31, 6.56599691e+07, 4.92345829e+41],
       [4.42413392e+05, 6.83767123e+30, 1.50609731e+35],
       [5.18470553e+21, 2.20264658e+04, 2.71828183e+00]])

In [53]:
# round/floor/ceil
np.random.random((2,3))*100

array([[39.33202357, 60.34415887, 44.18429507],
       [38.98373798, 43.59453489, 78.39581259]])

In [54]:
np.round(np.random.random((2,3))*100)

array([[96., 57., 97.],
       [68., 89., 47.]])

In [55]:
np.floor(np.random.random((2,3))*100)

array([[49., 85.,  9.],
       [ 6., 20., 82.]])

In [56]:
np.ceil(np.random.random((2,3))*100)

array([[74., 21., 43.],
       [97., 82., 97.]])

### Dimension

#### 0-D Array (Scalars)
* 0-D arrays, or Scalars, are the elements in an array. Each value in an array is a 0-D array.


In [57]:
zero_d_arr = np.array(255)
print(zero_d_arr)

255


#### 1-D Array
* An array that has 0-D arrays as its elements is called uni-dimensional or 1-D array.
* These are the most common and basic arrays.

In [58]:
one_d_arr = np.array([1,2,3,4,5])
print(one_d_arr)

[1 2 3 4 5]


#### 2-D Array (Matrix or 2nd Order Tensor)

* These are often used to represent matrix or 2nd order tensors.

In [59]:
two_d_arr = np.array([[1,2,3], [4,5,6]])
print(two_d_arr)

[[1 2 3]
 [4 5 6]]


#### 3-D Array (Matrices or 3rd Order Tensor)
* An array that has 2-D arrays (matrices) as its elements is called 3-D array.
* These are often used to represent a 3rd order tensor.
* 3-D arrays has two 2-D arrays.

#### Higher Dimension Array
* An array can have any number of dimensions.
* When the array is created, you can define the number of dimensions by using the ndmin argument.

In [60]:
higher_d_arr = np.array([1,2,3,4,5], ndmin = 5)
print(higher_d_arr)
print(f"Dimension of Array: {higher_d_arr.ndim}")

[[[[[1 2 3 4 5]]]]]
Dimension of Array: 5


### Numpy Array Indexing

#### Access 1-D Array

In [61]:
arr = np.array([1,2,3,4,5])
print(arr[2])
print(arr[0])

3
1


#### Access 2-D Array

In [62]:
arr = np.array([[1,2,3],[4,5,6]])
print(arr)

[[1 2 3]
 [4 5 6]]


In [63]:
print(f"second element of 1st row: {arr[0,1]}")

second element of 1st row: 2


In [64]:
print(f"first element of 1st row: {arr[0,0]}")

first element of 1st row: 1


In [65]:
print(f"sixth element of 2nd row: {arr[1,2]}")

sixth element of 2nd row: 6


In [66]:
print(f"Fifth element of 2nd row: {arr[1,1]}")

Fifth element of 2nd row: 5


#### Access 3-D Array

In [67]:
three_d_array = np.array([[[1,2,3],[4,5,6]], [[7,8,9],[10,11,12]]])
print(three_d_array)

[[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]]


In [68]:
print(three_d_array[0]) 
# The first number represents the first dimension, which contains two arrays:

[[1 2 3]
 [4 5 6]]


In [69]:
print(three_d_array[0,1])
# The second number represents the second dimension, which also contains two arrays:
# Since we selected 1, we are left with the second array:

[4 5 6]


In [70]:
print(three_d_array[0, 1, 2])
# The third number represents the third dimension, which contains three values:
# Since we selected 2, we end up with the third value:

6


In [71]:
print(three_d_array[1])

[[ 7  8  9]
 [10 11 12]]


In [72]:
print(three_d_array[1,0])

[7 8 9]


In [73]:
print(three_d_array[1,0,1])

8


In [74]:
print(three_d_array)

[[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]]


In [75]:
print(three_d_array[1,1,2])

12


In [76]:
print(three_d_array[1,0,0])

7


#### Negative Indexing

In [77]:
arr = np.array([[1,2,3,4,5],[6,7,8,9,10]])
print(f"The last element of an array: {arr[1,-1]}")

The last element of an array: 10


# NumPy Array Slicing

* Slicing means taking elements from one given index to another given index.
* Syntax:
    [start:end:step]
* Default value of `start` is `0` if did not define.
* Default value of `end` is `length` of the array.
* If we define value in `end` it is excluded and will give result `n-1`. (`n` means length of the array).
* Default value of `step` is `1` if did not define.
* When requirement is for particular `column` we requried all `rows`.
* When requirement is for particular `row` we required all `columns`.


In [78]:
arr = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
print(arr[1:9])

[2 3 4 5 6 7 8 9]


In [79]:
print(arr[4:])

[ 5  6  7  8  9 10]


In [80]:
print(arr[:4])

[1 2 3 4]


#### Negative Slicing

In [81]:
arr = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
print(arr[-3:-1])
# According to reule start value must be less than end value.
# In above example -3 < -1.

[8 9]


In [82]:
print(arr[-1:-3])
# Result is empty array because start value is greater than end value.

[]


In [83]:
print(arr[1:5:1])

[2 3 4 5]


In [84]:
print(arr[1:5:2])

[2 4]


In [85]:
print(arr[::2])

[1 3 5 7 9]


#### Slicing 2-D Arrays

In [86]:
# From the second element, slice elements from index 1 to index 4 (not included):
arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]])
print(arr[1, 1:4])

[7 8 9]


In [87]:
# From both elements, return index 2:
arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]])
print(arr)
print(arr[0:2, 2])

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]]
[3 8]


In [88]:
# From both elements, slice index 1 to index 4 (not included), 
# this will return a 2-D array:
arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]])
print(arr)

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]]


In [89]:
print(arr[0:2, 1:4])

[[2 3 4]
 [7 8 9]]


In [90]:
arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]])
print(arr)

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]]


In [91]:
print(arr[:,2:4])

[[3 4]
 [8 9]]


In [92]:
arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15]])
print(arr[1:, 3:])

[[ 9 10]
 [14 15]]


In [93]:
arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15]])
print(arr)

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]]


In [94]:
print(arr[::2, ::4])

[[ 1  5]
 [11 15]]


In [95]:
arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15]])
print(arr)

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]]


In [96]:
print(arr[::2,1:4:2])

[[ 2  4]
 [12 14]]


In [97]:
print(arr[1,0::4])

[ 6 10]


In [98]:
print(arr[0:2, 2::])

[[ 3  4  5]
 [ 8  9 10]]


In [99]:
print(arr[0:2, 2::2])

[[ 3  5]
 [ 8 10]]


#### Slicing 3-D Arrays

In [100]:
arr = np.arange(27).reshape(3,3,3)
print(arr)

[[[ 0  1  2]
  [ 3  4  5]
  [ 6  7  8]]

 [[ 9 10 11]
  [12 13 14]
  [15 16 17]]

 [[18 19 20]
  [21 22 23]
  [24 25 26]]]


In [101]:
# From the second element, slice elements from index 1 to index 4 (not included):
print(arr[1])

[[ 9 10 11]
 [12 13 14]
 [15 16 17]]


In [102]:
print(arr[::2])

[[[ 0  1  2]
  [ 3  4  5]
  [ 6  7  8]]

 [[18 19 20]
  [21 22 23]
  [24 25 26]]]


In [103]:
print(arr[0,1])

[3 4 5]


In [104]:
print(arr)

[[[ 0  1  2]
  [ 3  4  5]
  [ 6  7  8]]

 [[ 9 10 11]
  [12 13 14]
  [15 16 17]]

 [[18 19 20]
  [21 22 23]
  [24 25 26]]]


In [105]:
print(arr[1,::,1])

[10 13 16]


In [106]:
print(arr[2,1:,1:])

[[22 23]
 [25 26]]


In [107]:
print(arr)

[[[ 0  1  2]
  [ 3  4  5]
  [ 6  7  8]]

 [[ 9 10 11]
  [12 13 14]
  [15 16 17]]

 [[18 19 20]
  [21 22 23]
  [24 25 26]]]


In [108]:
print(arr[0::2, 0, ::2])

[[ 0  2]
 [18 20]]


# Iterating

In [109]:
one_d_array = np.arange(10).reshape(10)
two_d_array = np.arange(12).reshape(3,4)
three_d_array = np.arange(27).reshape(3,3,3)

In [110]:
# This will fetch all elements from 1-D array
for i in one_d_array:
    print(i)

0
1
2
3
4
5
6
7
8
9


In [111]:
# This will fetch one row at a time
for i in two_d_array:
    print(i)

[0 1 2 3]
[4 5 6 7]
[ 8  9 10 11]


In [112]:
# This will fetch one 2-D array at a time.
for i in three_d_array:
    print(i)

[[0 1 2]
 [3 4 5]
 [6 7 8]]
[[ 9 10 11]
 [12 13 14]
 [15 16 17]]
[[18 19 20]
 [21 22 23]
 [24 25 26]]


In [113]:
# This will fetch all the items in the array
# nditer convert 3-D array into 1-D array and then fetch all the elements in the array.
# This syntax will use for any dimension array.
for i in np.nditer(three_d_array):
    print(i, end=", ")

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 

# Reshaping
* This is a temporary operation.
* Whenever we see output it's means it is temporary

#### reshape()

#### transpose()

In [114]:
print(two_d_array,"\n")

print(np.transpose(two_d_array))

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]] 

[[ 0  4  8]
 [ 1  5  9]
 [ 2  6 10]
 [ 3  7 11]]


In [115]:
two_d_array.T # This is shortest way

array([[ 0,  4,  8],
       [ 1,  5,  9],
       [ 2,  6, 10],
       [ 3,  7, 11]])

#### ravel()

In [116]:
# This will convert any dimension array into 1-D array
print(two_d_array)
two_d_array.ravel()

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [117]:
three_d_array.ravel()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26])

 # Stacking
 * This is to be used for join two arrays.
 * The shape of the arrays must be the same.
 * Any number of arrays.
 * This could be horizontal.
     *  If we have 2 arrays `[ 2 x 2 ]` `[ 2 x 2 ]` it will convert into `[ 2 x 4 ]`
 * This could be vertical.
     * If we have 2 arrays `[ 2 x 2 ]` `[ 2 x 2 ]` it will convert into `[ 4 x 2 ]`
  

#### Horizontal Stacking

In [118]:
a4 = np.arange(12).reshape(3,4)
a5 = np.arange(12, 24).reshape(3,4)

In [119]:
np.hstack((a4, a5))

array([[ 0,  1,  2,  3, 12, 13, 14, 15],
       [ 4,  5,  6,  7, 16, 17, 18, 19],
       [ 8,  9, 10, 11, 20, 21, 22, 23]])

In [120]:
np.hstack((a4, a5, two_d_array))

array([[ 0,  1,  2,  3, 12, 13, 14, 15,  0,  1,  2,  3],
       [ 4,  5,  6,  7, 16, 17, 18, 19,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 20, 21, 22, 23,  8,  9, 10, 11]])

In [121]:
np.hstack((a4, a5, two_d_array))

array([[ 0,  1,  2,  3, 12, 13, 14, 15,  0,  1,  2,  3],
       [ 4,  5,  6,  7, 16, 17, 18, 19,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 20, 21, 22, 23,  8,  9, 10, 11]])

#### Vertical Stacking

In [122]:
np.vstack((a4, a5))

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

# Splitting
* This is just an opposite of `stacking`.
* It has to be equal division. 


#### Horizontal Spliting

In [123]:
two_d_array

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [124]:
np.hsplit(two_d_array, 2)

[array([[0, 1],
        [4, 5],
        [8, 9]]),
 array([[ 2,  3],
        [ 6,  7],
        [10, 11]])]

In [125]:
np.hsplit(two_d_array, 4)

[array([[0],
        [4],
        [8]]),
 array([[1],
        [5],
        [9]]),
 array([[ 2],
        [ 6],
        [10]]),
 array([[ 3],
        [ 7],
        [11]])]

In [126]:
# np.hsplit(two_d_array, 5) # This would be error

#### Vertical Spliting

In [127]:
two_d_array

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [128]:
np.vsplit(two_d_array, 3)

[array([[0, 1, 2, 3]]), array([[4, 5, 6, 7]]), array([[ 8,  9, 10, 11]])]

In [129]:
# np.vsplit(two_d_array, 2) # This would be error

# NumPy Array Copy vs View

 * The main difference between a copy and a view of an array is that the copy is a new array, and the view is just a view of the original array.
 * The copy owns the data and any changes made to the copy will not affect original array, and any changes made to the original array will not affect the copy.
 * The view does not own the data and any changes made to the view will affect the original array, and any changes made to the original array will affect the view.

In [130]:
import numpy as np

arr = np.array([1, 2, 3, 4, 5])
x = arr.copy()
arr[0] = 42

print(arr)
print(x)

[42  2  3  4  5]
[1 2 3 4 5]


In [131]:
import numpy as np

arr = np.array([1, 2, 3, 4, 5])
x = arr.view()
arr[0] = 42

print(arr)
print(x)

[42  2  3  4  5]
[42  2  3  4  5]


In [132]:
# Print the value of the base attribute to check if an array owns it's data or not:
import numpy as np

arr = np.array([1, 2, 3, 4, 5])

x = arr.copy()
y = arr.view()

print(x.base)
print(y.base)

None
[1 2 3 4 5]


# Extra Advance Spice

#### unravel_index
The `numpy.unravel_index` function in NumPy is used to convert a flat index or indices into an array of shape `shape` into a tuple of coordinate arrays. In simpler terms, it helps you find the multi-dimensional indices corresponding to a given flat index in a multi-dimensional array.

Here is the basic syntax:

```Python
numpy.unravel_index(indices, shape, order='C')
```
- `indices`: An integer or array of integers representing the flat index or indices to be converted.
- `shape`: The shape of the array.
- `order`: Specifies whether to index the array using C-style (row-major) or Fortran-style (column-major) indexing. The default is 'C'.

Let's look at a simple example:

In this example, the array has a shape of (3, 4, 5), and we want to find the indices corresponding to the flat index 17. The output will be:

```
Indices: (0, 3, 2)
```

This means that the element at the flat index 17 in an array of shape (3, 4, 5) corresponds to the element at position (0, 3, 2). The order of the indices corresponds to the dimensions of the array.

This function is particularly useful when working with flattened indices, such as when you're dealing with a 1D representation of a multi-dimensional array.

In [133]:
import numpy as np

shape = (3, 4, 5)
flat_index = 17

indices = np.unravel_index(flat_index, shape)

print("Indices:", indices)

Indices: (0, 3, 2)


#### linalg.norm

`numpy.linalg.norm` is a function in NumPy that computes one of several types of matrix norms or vector norms. It is a part of the NumPy linear algebra module (`numpy.linalg`).

Here is the basic syntax:

```python
numpy.linalg.norm(x, ord=None, axis=None, keepdims=False)
```

- `x`: The input array (can be a vector or matrix).
- `ord`: The order of the norm. It can be one of the following:
  - For vectors:
    - `ord=None` or `'fro'`: Frobenius norm (default if `x` is a matrix).
    - `ord=1`: L1 norm (sum of absolute values).
    - `ord=2`: L2 norm (Euclidean norm).
    - `ord=np.inf`: Infinity norm (maximum absolute value).
    - `ord=-np.inf`: Negative infinity norm (minimum absolute value).
  - For matrices, additional norms are available.
- `axis`: If specified, it computes the norm along the specified axis or axes of the array.
- `keepdims`: If True, the result will broadcast correctly against the input array.

Here are a few examples:

```python
import numpy as np

# Example 1: L2 norm of a vector
v = np.array([3, 4])
l2_norm = np.linalg.norm(v)
print("L2 Norm:", l2_norm)

# Example 2: Frobenius norm of a matrix
A = np.array([[1, 2], [3, 4]])
frobenius_norm = np.linalg.norm(A)
print("Frobenius Norm:", frobenius_norm)

# Example 3: L1 norm along a specific axis
B = np.array([[1, -2, 3], [4, 5, -6]])
l1_norm_along_axis = np.linalg.norm(B, ord=1, axis=1)
print("L1 Norm along axis 1:", l1_norm_along_axis)
```

In the first example, it calculates the L2 norm of a vector. In the second example, it calculates the Frobenius norm of a matrix. In the third example, it computes the L1 norm along axis 1 for a matrix.

These norms are essential in various mathematical and statistical applications, providing a measure of the size or magnitude of vectors or matrices.

In [134]:
import numpy as np

# Example 1: L2 norm of a vector
v = np.array([3, 4])
l2_norm = np.linalg.norm(v)
print("L2 Norm:", l2_norm)

# Example 2: Frobenius norm of a matrix
A = np.array([[1, 2], [3, 4]])
frobenius_norm = np.linalg.norm(A)
print("Frobenius Norm:", frobenius_norm)

# Example 3: L1 norm along a specific axis
B = np.array([[1, -2, 3], [4, 5, -6]])
l1_norm_along_axis = np.linalg.norm(B, ord=1, axis=1)
print("L1 Norm along axis 1:", l1_norm_along_axis)


L2 Norm: 5.0
Frobenius Norm: 5.477225575051661
L1 Norm along axis 1: [ 6. 15.]


#### exp

`numpy.exp` is a NumPy function that calculates the exponential of each element in an input array. The exponential function is commonly used in mathematical and scientific computations. The formula for the exponential function is \(e^x\), where \(e\) is Euler's number (approximately 2.71828), and \(x\) is the input value.

Here's the basic syntax of `numpy.exp`:

```python
numpy.exp(x, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature, extobj])
```

- `x`: The input array.
- `out`: If provided, the result will be placed in this array. It must have the same shape as the input. If not provided or `None`, a new array is created.
- `where`: A boolean array of the same shape as `x`, determining where to calculate the result.
- `dtype`: The desired data-type for the array. If not given, infer the data type from the input data.

Here's a simple example:

```python
import numpy as np

x = np.array([1.0, 2.0, 3.0])

# Calculate the exponential of each element
exp_values = np.exp(x)

print("Original Array:")
print(x)
print("\nExponential Values:")
print(exp_values)
```

In this example, `np.exp(x)` returns a new array where each element is the exponential of the corresponding element in the original array `x`. The output will be:

```
Original Array:
[1. 2. 3.]

Exponential Values:
[ 2.71828183  7.3890561  20.08553692]
```

The `numpy.exp` function is particularly useful in various mathematical and statistical applications where exponential transformations are needed.

# Advance Numpy

## Numpy Array vs Pyhton List

### Speed

In [135]:
a = [i for i in range(10000000)]
b = [i for i in range(10000000, 20000000)]
c = []
import time
start = time.time()

for i in range(len(a)):
    c.append(a[i] + b[i])
print(time.time()-start)

5.713846206665039


In [136]:
import numpy as np

a = np.arange(10000000)
b = np.arange(10000000, 20000000)
import time
start = time.time()
c = a + b
print(time.time()-start)

0.17889189720153809


In [137]:
4.06/0.05

81.19999999999999

### Memory

In [138]:
a = [i for i in range(10000000)]
import sys

print("Memory occupied by Python list {} bytes in memory".format(sys.getsizeof(a)))


Memory occupied by Python list 89095160 bytes in memory


In [139]:
a = np.arange(10000000)
import sys

print("Memory occupied by Numpy Array {} bytes in memory".format(sys.getsizeof(a)))

Memory occupied by Numpy Array 40000104 bytes in memory


# Advance Indexing

### Fancing Indexing

In [153]:
a = np.arange(24).reshape(6,4)
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

In [154]:
a[[0,2,3,5]]

array([[ 0,  1,  2,  3],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [20, 21, 22, 23]])

In [155]:
a[:,[0,2,3]]

array([[ 0,  2,  3],
       [ 4,  6,  7],
       [ 8, 10, 11],
       [12, 14, 15],
       [16, 18, 19],
       [20, 22, 23]])

In [158]:
a[[0,1],[0,2]]

array([0, 6])

### Boolean Indexing
* Filtering the data based on given condition

In [143]:
a = np.random.randint(1, 100, 24).reshape(6,4)
a

array([[66, 31, 84,  4],
       [46, 36, 88, 50],
       [51, 93, 99, 70],
       [52, 75, 88, 72],
       [51, 82, 82, 40],
       [95, 13, 48, 60]])

In [144]:
# Find all numbers greater than 50
a > 50
# Result would be boolean array

array([[ True, False,  True, False],
       [False, False,  True, False],
       [ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True,  True,  True, False],
       [ True, False, False,  True]])

In [145]:
a[a>50]

array([66, 84, 88, 51, 93, 99, 70, 52, 75, 88, 72, 51, 82, 82, 95, 60])

In [146]:
# Find out even numbers
a[a%2==0]

array([66, 84,  4, 46, 36, 88, 50, 70, 52, 88, 72, 82, 82, 40, 48, 60])

In [147]:
# find all numbers greater than 50 and are even

a[(a>50) & (a % 2 == 0)] # Use bitwise operator '&' instead of logical operator 'and'

array([66, 84, 88, 70, 52, 88, 72, 82, 82, 60])

In [148]:
# find all number not divisible by 7
a[~(a % 7 == 0)] # Use bitwise operator '~' instead of 'Not'

array([66, 31,  4, 46, 36, 88, 50, 51, 93, 99, 52, 75, 88, 72, 51, 82, 82,
       40, 95, 13, 48, 60])

# Broadcasting